unmarshaler.go

```diff
@@ -377,7 +377,11 @@ func (d *decoder) handleArrayTableCollection(key unstable.Iterator, v reflect.Va
 
 		return v, nil
 	case reflect.Slice:
-		elem := v.Index(v.Len() - 1)
+		idx := v.Len() - 1
+		if idx < 0 {
```

Comment:
This seems odd to me. I'd expect something like a type check, so that we can confidentially return an error along the lines of "you cannot put an array of tables into something that is not a slice or an array" (there may be better wording 😅 )

Reply:
IIUC this index panic must be dealt with first. Perhaps my original example was not minimal enough. Consider this version, with the only change being an explicit `[[A]]` before `[[A.B]]`:
https://go.dev/play/p/rJqxTrNgONH
> toml: cannot decode TOML array table into struct field struct { B struct {} }.B of type struct {}

Should the slice be expanded instead, so that we proceed along to this error ^?

Reply:
Implicitly, `A` is a table. Adding an explicit `[[A]]` actually changes the structure of the document to make it an array of tables, so I don't think we can compare those.

To me, the logical path of the unmarshaler is:

1. Is `A` a key in the current target? -> ✅ it's a field in the root struct. That value becomes the current target.
2. Is `B` a key in the current target? -> ❎ error: the current target is a slice, so B not a key; and `A` hasn't been marked as an array table.

The problem is, the unmarshaler doesn't explicitly tracks which tables are array tables or not for the sake of performance (except in strict mode). However, when it encounters an explicit array table, it guarantees that the slices have been expended as needed to be ready to receive a key lookup. This is because `[[A]]` is equivalent to `{"A": [{}]}` in json: https://go.dev/play/p/74TSvAtr8_R.

So, we should be able to use the fact that the unmarshaler is finding itself at a point where the container is a slice but hasn't been allocated as an indicator that the type are incompatibles, and document is possibly missing an explicit array table.